src/pybind/mgr/drain_osd/module.py

```diff
@@ -0,0 +1,329 @@
+        self.run = False
+        self.event.set()
+
+    def osds_not_in_cluster(self, osd_ids: List[int]) -> Set[int]:
+        self.log.info(f"Checking if provided osds <{osd_ids}> exist in the cluster")
+        osd_map = self.get_osdmap()
+        cluster_osds = [x.get('osd') for x in osd_map.dump().get('osds', [])]
+        not_in_cluster = set()
+        for osd_id in osd_ids:
+            if int(osd_id) not in cluster_osds:
+                self.log.error(f"Could not find {osd_id} in cluster")
+                not_in_cluster.add(osd_id)
+        return not_in_cluster
+
+    def empty_osds(self, osd_ids: Set[int]) -> List[int]:
+        if not osd_ids:
+            return list()
+        osd_df_data = self.osd_df()
+        empty_osds = list()
+        for osd_id in osd_ids:
+            if self.is_empty(osd_id, osd_df=osd_df_data):
+                empty_osds.append(osd_id)
+        return empty_osds
+
+    def osd_df(self) -> dict:
+        # TODO: this should be cached I think
+        base_cmd = 'osd df'
+        ret, out, err = self.mon_command({
+            'prefix': base_cmd,
+            'format': 'json'
+        })
+        return json.loads(out)
+
+    def is_empty(self, osd_id: int, osd_df: Optional[dict] = None) -> bool:
+        pgs = self.get_pg_count(osd_id, osd_df=osd_df)
+        if pgs != 0:
+            self.log.info(f"osd: {osd_id} still has {pgs} PGs.")
+            return False
+        self.log.info(f"osd: {osd_id} has no PGs anymore")
+        return True
+
+    def reweight_osds(self, osd_ids: Set[int]) -> bool:
+        results = [(self.reweight_osd(osd_id)) for osd_id in osd_ids]
+        return all(results)
+
+    def get_pg_count(self, osd_id: int, osd_df: Optional[dict] = None) -> int:
+        if not osd_df:
+            osd_df = self.osd_df()
+        osd_nodes = osd_df.get('nodes', [])
+        for osd_node in osd_nodes:
+            if osd_node.get('id', None) == int(osd_id):
+                return osd_node.get('pgs')
+        errmsg = f"Could not find <pgs> field for osd_id: {osd_id} in osd_df data"
+        self.log.error(errmsg)
+        raise RuntimeError(errmsg)
+
+    def get_osd_weight(self, osd_id: int) -> float:
+        osd_df = self.osd_df()
+        osd_nodes = osd_df.get('nodes', [])
+        for osd_node in osd_nodes:
+            if osd_node.get('id', None) == int(osd_id):
+                return float(osd_node.get('crush_weight'))
+        return -1.0
+
+    def reweight_osd(self, osd_id: int, weight: float = 0.0) -> bool:
+        if self.get_osd_weight(osd_id) == weight:
+            self.log.debug(f"OSD <{osd_id}> is already weighted with: {weight}")
+            return True
+        base_cmd = 'osd crush reweight'
+        ret, out, err = self.mon_command({
+            'prefix': base_cmd,
+            'name': f'osd.{osd_id}',
+            'weight': weight
+        })
+        cmd = f"{base_cmd} on osd.{osd_id} to weight: {weight}"
+        self.log.debug(f"running cmd: {cmd}")
```

Comment:
It's better to place this log before the `self.mon_command()`, and setting it to `info` level.